File: python/tests/test_ops.py
import numpy as np
import torch
import torch.nn.functional as F
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from ops import softmax, gelu, layer_norm, scaled_dot_product_attention, multi_head_attention

# Tolerance for float32 comparisons
ATOL = 1e-5
RTOL =1e-5

def assert_close(a: np.ndarray, b: np.ndarray, name: str) -> None:
    if not np.allclose(a, b, atol=ATOL, rtol=RTOL):
        max_diff = np.abs(a - b).max()
        raise AssertionError(f"FAIL [{name}] max_diff={max_diff:.2e}")
    print(f"  PASS [{name}]")


def test_softmax():
    print("softmax")
    rng = np.random.default_rng(0)

    # Standard case
    x = rng.standard_normal((4, 16)).astype(np.float32)
    ref = F.softmax(torch.tensor(x), dim=-1).numpy()
    assert_close(softmax(x), ref, "random_input")

    # Large values - numerical stability check
    x_large = x * 100.0
    ref_large = F.softmax(torch.tensor(x_large), dim=-1).numpy()
    assert_close(softmax(x_large), ref_large, "large values (stability)")

    # Rows sum to 1
    out = softmax(x)
    row_sums = out.sum(axis=-1)
    assert np.allclose(row_sums, 1.0, atol=1e-6), f"rows don't sum to 1: {row_sums}"
    print("  PASS [rows sum to 1.0]")


def test_gelu():
    print("gelu")
    rng = np.random.default_rng(1)
    x = rng.standard_normal((8, 32)).astype(np.float32)
    ref = F.gelu(torch.tensor(x), approximate="tanh").numpy()
    assert_close(gelu(x), ref, "random input")

    # Known values
    assert_close(gelu(np.array([0.0])), np.array([0.0]), "gelu(0) = 0")
    # gelu is approximately linear for large positive x
    val = gelu(np.array([10.0]))[0]
    assert abs(val - 10.0) < 0.01, f"gelu(10) should be ~10, got {val}"
    print("  PASS [known values]")


def test_layer_norm():
    print("layer_norm")
    rng = np.random.default_rng(2)
    T, D = 6, 64

    x     = rng.standard_normal((T, D)).astype(np.float32)
    gamma = rng.standard_normal((D,)).astype(np.float32)
    beta  = rng.standard_normal((D,)).astype(np.float32)

    # PyTorch reference
    ln = torch.nn.LayerNorm(D, elementwise_affine=True, eps=1e-5)
    ln.weight.data = torch.tensor(gamma)
    ln.bias.data   = torch.tensor(beta)
    ref = ln(torch.tensor(x)).detach().numpy()

    assert_close(layer_norm(x, gamma, beta), ref, "random input")

    # With identity gamma/beta, output should have mean~0 and std~1 per row
    ones  = np.ones(D, dtype=np.float32)
    zeros = np.zeros(D, dtype=np.float32)
    out   = layer_norm(x, ones, zeros)
    row_means = out.mean(axis=-1)
    row_stds  = out.std(axis=-1)
    assert np.allclose(row_means, 0.0, atol=1e-5), f"means not 0: {row_means}"
    assert np.allclose(row_stds,  1.0, atol=1e-3), f"stds not 0: {row_stds}"
    print("  PASS [identity gamma/beta: mean≈0, std≈1]")

def test_scaled_dot_product_attention():
    print("scaled_dot_product_attention")
    rng = np.random.default_rng(3)
    T, d_k, d_v = 5, 16, 16
    
    Q = rng.standard_normal((T, d_k)).astype(np.float32)
    K = rng.standard_normal((T, d_k)).astype(np.float32)
    V = rng.standard_normal((T, d_v)).astype(np.float32)

    out, weights = scaled_dot_product_attention(Q, K, V)

    # Shape checks
    assert out.shape     == (T, d_v), f"outshape {out.shape}"
    assert weights.shape == (T, T),   f"weights shape {weights.shape}"
    print("  PASS [output shapes]")

    # Weights should be a valid probability distribution
    assert np.allclose(weights.sum(axis=-1), 1.0, atol=1e-5), "weights don't sum to 1"
    print("  PASS [attention weights sum to 1]")

    #With causal mask: upper triangle of weights should be ~0
    from ops import causal_mask
    mask = causal_mask(T)
    _, masked_weights = scaled_dot_product_attention(Q, K, V, mask)
    upper = masked_weights[np.triu_indices(T, k=1)]
    assert np.allclose(upper, 0.0, atol=1e-6), f"causal mask leak: max={upper.max():.2e}"
    print("  PASS [causal mask: upper triangle ≈ 0]")

def test_multi_head_attention():
    print("multi_head_attention")
    rng = np.random.default_rng(4)
    T, d_model, n_head = 5, 64, 4
    
    x        = rng.standard_normal((T, d_model)).astype(np.float32)
    c_attn_w = rng.standard_normal((d_model, 3 * d_model)).astype(np.float32)
    c_attn_b = rng.standard_normal((3 * d_model,)).astype(np.float32)
    c_proj_w = rng.standard_normal((d_model, d_model)).astype(np.float32)
    c_proj_b = rng.standard_normal((d_model,)).astype(np.float32)

    from ops import causal_mask
    mask = causal_mask(T)
    out, weights = multi_head_attention(
        x, c_attn_w, c_attn_b, c_proj_w, c_proj_b, n_head, mask
    )

    assert out.shape     == (T, d_model),       f"out shape {out.shape}"
    assert weights.shape == (n_head, T, T), f"weights shape {weights.shape}"
    print("  PASS [output shapes]")

    # Per-head attention weights sum to 1 along key dimension
    sums = weights.sum(axis=-1)  # [n_head, T]
    assert np.allclose(sums, 1.0, atol=1e-5), "per-head weights don't sum to 1"
    print("  PASS [per-head weights sum to 1]")


if __name__ == "__main__":
    print("=" * 50)
    print("ops.py unit tests")
    print("=" * 50)
    test_softmax()
    test_gelu()
    test_layer_norm()
    test_scaled_dot_product_attention()
    test_multi_head_attention()
    print("=" * 50)
    print("All tests passed.")

File: cpp/tests/test_ops.cpp
// cpp/tests/test_ops.cpp
#include "ops.hpp"
#include <iostream>
#include <vector>
#include <cmath>
#include <cassert>
#include <numeric>

static bool close(float a, float b, float atol = 1e-4f)
{
    return std::abs(a - b) <= atol;
}

static void assert_close_vec(
    const std::vector<float> &got,
    const std::vector<float> &ref,
    const std::string &name,
    float atol = 1e-4f)
{
    assert(got.size() == ref.size());
    float max_diff = 0.0f;
    for (size_t i = 0; i < got.size(); i++)
    {
        max_diff = std::max(max_diff, std::abs(got[i] - ref[i]));
    }
    if (max_diff > atol)
    {
        std::cerr << "FAIL [" << name << "] max_diff=" << max_diff << "\n";
        std::exit(1);
    }
    std::cout << "  PASS [" << name << "] max_diff=" << max_diff << "\n";
}

void test_softmax()
{
    std::cout << "softmax\n";

    // Row 1: [1, 2, 3] Row 2: [1, 1, 1]
    std::vector<float> x = {1.0f, 2.0f, 3.0f,
                            1.0f, 1.0f, 1.0f};
    ops::softmax(x.data(), 2, 3);

    // Row 1 reference (computed from NumPy)
    // softmax([1,2,3]) = [0.0900, 0.2447, 0.6652]
    assert(close(x[0], 0.0900f, 1e-3f));
    assert(close(x[1], 0.2447f, 1e-3f));
    assert(close(x[2], 0.6652f, 1e-3f));
    std::cout << "  PASS [known values]\n";

    // Row 2: uniform input -> uniform output
    assert(close(x[3], 1.0f / 3.0f, 1e-5f));
    assert(close(x[4], 1.0f / 3.0f, 1e-5f));
    assert(close(x[5], 1.0f / 3.0f, 1e-5f));
    std::cout << "  PASS [uniform input]\n";

    // Rows sum to 1
    float sum1 = x[0] + x[1] + x[2];
    float sum2 = x[3] + x[4] + x[5];
    assert(close(sum1, 1.0f, 1e-5f));
    assert(close(sum2, 1.0f, 1e-5f));
    std::cout << "  PASS [rows sum to 1]\n";
}

void test_gelu()
{
    std::cout << "gelu\n";

    std::vector<float> x = {0.0f, 1.0f, -1.0f, 2.0f};
    ops::gelu(x.data(), 4);

    // gelu(0) = 0
    assert(close(x[0], 0.0f, 1e-6f));
    std::cout << "  PASS [gelu(0)]\n";

    // gelu(1) ≈ 0.8413
    assert(close(x[1], 0.8413f, 1e-3f));
    std::cout << "  PASS [gelu(1)≈0.8413]\n";

    // gelu(-1) ≈ -0.1587
    assert(close(x[2], -0.1587f, 1e-3f));
    std::cout << "  PASS [gelu(-1)≈-0.1587]\n";
}

void test_layer_norm()
{
    std::cout << "layer_norm\n";
    int T = 2, D = 4;

    std::vector<float> x = {1.0f, 2.0f, 3.0f, 4.0f,
                            2.0f, 2.0f, 2.0f, 2.0f};
    std::vector<float> gamma(D, 1.0f);
    std::vector<float> beta(D, 0.0f);
    std::vector<float> out(T * D);

    ops::layer_norm(x.data(), out.data(), gamma.data(), beta.data(), T, D);

    // Row 2 is constant — after layernorm should be all zeros
    assert(close(out[4], 0.0f, 1e-5f));
    assert(close(out[5], 0.0f, 1e-5f));
    assert(close(out[6], 0.0f, 1e-5f));
    assert(close(out[7], 0.0f, 1e-5f));
    std::cout << "  PASS [constant row -> zeros]\n";

    // Row 1 should have mean≈0, std≈1
    float mean = (out[0] + out[1] + out[2] + out[3]) / 4.0f;
    assert(close(mean, 0.0f, 1e-5f));
    std::cout << "  PASS [mean≈0]\n";
}

void test_causal_mask()
{
    std::cout << "causal_mask\n";
    int T = 4;
    std::vector<float> mask(T * T);
    ops::causal_mask(mask.data(), T);

    // Diagonal and below: 0.0
    for (int i = 0; i < T; i++)
    {
        for (int j = 0; j <= i; j++)
        {
            assert(close(mask[i * T + j], 0.0f, 1e-6f));
        }
    }
    std::cout << "  PASS [lower triangle = 0]\n";

    // Above diagonal: large negative
    for (int i = 0; i < T; i++)
    {
        for (int j = i + 1; j < T; j++)
        {
            assert(mask[i * T + j] < -1e9f);
        }
    }
    std::cout << "  PASS [upper triangle = -inf]\n";
}

void test_matmul()
{
    std::cout << "matmul\n";

    // 2x3 @ 3x2 = 2x2
    std::vector<float> A = {1, 2, 3, 4, 5, 6};
    std::vector<float> B = {7, 8, 9, 10, 11, 12};
    std::vector<float> C(4);

    ops::matmul(A.data(), B.data(), C.data(), 2, 3, 2);

    // [1*7+2*9+3*11, 1*8+2*10+3*12] = [58, 64]
    // [4*7+5*9+6*11, 4*8+5*10+6*12] = [139, 154]
    assert(close(C[0], 58.0f));
    assert(close(C[1], 64.0f));
    assert(close(C[2], 139.0f));
    assert(close(C[3], 154.0f));
    std::cout << "  PASS [2x3 @ 3x2 = 2x2]\n";
}

int main()
{
    std::cout << std::string(50, '=') << "\n";
    std::cout << "ops.cpp unit tests\n";
    std::cout << std::string(50, '=') << "\n";

    test_matmul();
    test_softmax();
    test_gelu();
    test_layer_norm();
    test_causal_mask();

    std::cout << std::string(50, '=') << "\n";
    std::cout << "All tests passed.\n";
    return 0;
};